integrations/servicenow/client.py

```diff
@@ -0,0 +1,103 @@
+from typing import Any, AsyncGenerator
+import base64
+import httpx
+from loguru import logger
+from enum import StrEnum
+from port_ocean.utils import http_async_client
+
+PAGE_SIZE = 100
+
+
+class ObjectKind(StrEnum):
+    GROUP = "sys_user_group"
+    CATALOG = "sc_catalog"
+    INCIDENT = "incident"
+
+
+class ServicenowClient:
+    def __init__(
+        self, servicenow_host: str, servicenow_username: str, servicenow_password: str
+    ):
+        self.servicenow_host = servicenow_host
+        self.servicenow_username = servicenow_username
+        self.servicenow_password = servicenow_password
+        self.http_client = http_async_client
+        self.http_client.headers.update(self.api_auth_params["headers"])
+
+    @property
+    def api_auth_params(self) -> dict[str, Any]:
+        auth_message = f"{self.servicenow_username}:{self.servicenow_password}"
+        auth_bytes = auth_message.encode("ascii")
+        b64_bytes = base64.standard_b64encode(auth_bytes)
+        b64_message = b64_bytes.decode("ascii")
+
+        return {
+            "headers": {
+                "Authorization": f"Basic {b64_message}",
+                "Content-Type": "application/json",
+            },
+        }
+
+    async def get_paginated_resource(
+        self, resource_kind: ObjectKind
+    ) -> AsyncGenerator[list[dict[str, Any]], None]:
+        offset = 0
+
+        while True:
+            params: dict[str, Any] = {
+                "sysparm_offset": offset * PAGE_SIZE,
+                "sysparm_limit": PAGE_SIZE,
+                "sysparm_query": "ORDERBYsys_created_on",
+            }
+            try:
+                response = await self.http_client.get(
+                    url=f"{self.servicenow_host}/api/now/table/{resource_kind}",
+                    params=params,
+                )
+                response.raise_for_status()
+                records = response.json().get("result", [])
+
+                if not records:
+                    break
+
+                total_rows = int(response.headers.get("X-Total-Count", 0))
```

Comment:
You override it in each query? can it change while you paginate?

Reply:
This is the total number of records in the kind so it is static

Reply:
I am afraid from raise condition, where the total count change while we query and therefor can cause infinite loop. Most of the times pagination api's return some link or id that you use to query the next page so that the API understands that you keep paginating as part of the first request context. 

Let's make sure that we don't miss something.

Reply:
Looking in their [docs](https://developer.servicenow.com/dev.do#!/reference/api/sandiego/rest/c_TableAPI%23table-GET?navFilter=table).

Seems like as part of the headers response you also get a link header in which it points you to the next request you need to perform.  I think we need to use this native approach, that way you just keep looping until there is no next link to go over or something like that.

<img width="993" alt="Screen Shot 2023-12-20 at 14 09 10" src="https://github.com/port-labs/ocean/assets/59158507/5f37a92b-f528-4ef6-b182-236c3d066756">